### ISSUE TITLE
Use shell redirections instead of "dd"

### ISSUE THREAD

**Issue by ksobolew** [digint/btrbk#105]
I finally was able to give btrbk a try (thanks!) but I noticed that the performance is really, really low. The CPU does barely notice it despite enabling compression and encryption. Then I noticed that the final output is written using dd.

Why?

I'm writing to a SMB network share (no ssh for me) and it manages to write about 200 kB/s max. With the default value of the block size dd is slow even in "normal" circumstances (e.g. writing to an USB drive). Can't this be changed somehow?

**Comment by digint** [comment]
This is work in progress. I originally used "dd" because redirections were not supported by my run_cmd() framework, which was a bit hacky and causes trouble on special mounts like fuse, SMB, NFS.

This is now  fixed/improved (in a019d8a3e37391cbedaf1d7ad43530ce4097500a) while implementing `stream_compress` functionality, which I just merged into the [devel branch](/digint/btrbk/tree/devel).

Give it a try, it should already work fine (see known bugs in ChangeLog)

**Comment by tobig** [comment]
As mentioned in PR100, the default block sizes of dd, but also the block sizes used by ">" seem to cause trouble with a acd_cli mount.

Let me provide you with some data:

## System information

$ stat -f /media/acd/
  File: "/media/acd/"
    ID: 0        Namelen: 256     Type: fuseblk
Block size: 524288     Fundamental block size: 524288
Blocks: Total: 209715200  Free: 209674111  Available: 209674111
Inodes: Total: 0          Free: 0

$ ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 30521
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 30521
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited

## Current performance (with stream_compress branch):


time 'btrfs' 'send' '-v' '/mnt/btrfs/btrbk_snapshots/test.20160822T1550' | 'gpg' '--batch' '--no-tty' '--trust-model' 'always' '--compress-algo' 'none' '--default-recipient' 'XXXXXXX '--encrypt' > /media/acd/btrbk/test/test.20160822T1550--be8d1a43-d805-0448-808e-b576e9439c54.btrfs.gpg.partc
At subvol /mnt/btrfs/btrbk_snapshots/test.20160822T1550
BTRFS_IOC_SEND returned 0
joining genl thread

real	0m59.123s
user	0m11.584s
sys	0m4.712s

## Using dd with 512K block size

 time 'btrfs' 'send' '-v' '/mnt/btrfs/btrbk_snapshots/test.20160822T1550' | 'gpg' '--batch' '--no-tty' '--trust-model' 'always' '--compress-algo' 'none' '--default-recipient' 'XXXXXX' '--encrypt' | dd bs=512K > /media/acd/btrbk/test/test.20160822T1550--be8d1a43-d805-0448-808e-b576e9439c54.btrfs.gpg.partc
At subvol /mnt/btrfs/btrbk_snapshots/test.20160822T1550
BTRFS_IOC_SEND returned 0
joining genl thread
0+19195 records in
0+19195 records out
765762358 bytes (766 MB, 730 MiB) copied, 8.13897 s, 94.1 MB/s

real	0m8.753s
user	0m7.512s
sys	0m1.972s

Using dd gives a 7.2x increase in performance.

# Use lz4 to reach a similar speedup:

time 'btrfs' 'send' '-v' '/mnt/btrfs/btrbk_snapshots/test.20160822T1550' | 'gpg' '--batch' '--no-tty' '--trust-model' 'always' '--compress-algo' 'none' '--default-recipient' 'XXXXXXX' '--encrypt' | lz4 -c > /media/acd/btrbk/test/test.20160822T1550--be8d1a43-d805-0448-808e-b576e9439c54.btrfs.gpg.partc
At subvol /mnt/btrfs/btrbk_snapshots/test.20160822T1550
BTRFS_IOC_SEND returned 0
joining genl thread

real	0m14.997s
user	0m8.032s
sys	0m2.172s

It seems lz4 will also ensure that writes happen in larger chunks.

# Use cat to reach similar performance

time 'btrfs' 'send' '-v' '/mnt/btrfs/btrbk_snapshots/test.20160822T1550' | 'gpg' '--batch' '--no-tty' '--trust-model' 'always' '--compress-algo' 'none' '--default-recipient' 'XXXXXXX' '--encrypt' | cat -  > /media/acd/btrbk/test/test.20160822T1550--be8d1a43-d805-0448-808e-b576e9439c54.btrfs.gpg.partc
At subvol /mnt/btrfs/btrbk_snapshots/test.20160822T1550
BTRFS_IOC_SEND returned 0
joining genl thread

real	0m10.244s
user	0m7.648s
sys	0m2.004s

# Conclusion 

Redirecting gpg output directly to the file system does not seem to use the filesystem recommendation for optimal write block sizes, but likely uses smaller blocks. If we increase the block size using dd (or just other do-nothing commands), this issue can be worked around.


PS: I earlier had a patch which used 4MByte blocks, which also reached similar performance, but is clearly larger than necessary.

**Comment by digint** [comment]
I got curious about write sizes of `cat` and `gpg`, this is what I found:

### Test write size of "cat"
```
truncate -s 1M /tmp/infile
echo 'exec cat /tmp/infile > /tmp/out' > /tmp/test_cat.sh
strace -f sh ./test_cat.sh
[...]
read(3, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 131072) = 131072
write(1, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 131072) = 131072
read(3, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 131072) = 131072
write(1, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 131072) = 131072
[...]
```

Hmm. So `cat` writes in chunks of 131072 == 128k. Why? Looking at the code helps:

[outsize](/coreutils/coreutils/blob/v8.25/src/cat.c#L639) is defined by the [io_blksize macro](/coreutils/coreutils/blob/v8.25/src/ioblksize.h#L75): `max(128*1024, st_blksize)`, and because tmpfs has a blksize=4096 we get 128k for the writes.

### Now the same for "gpg":

```
truncate -s 1M /tmp/infile
echo 'exec gpg -e --compress-algo none --default-recipient axel@tty0.ch < /tmp/infile > /tmp/out' > /tmp/test_gpg.sh
strace -f sh ./test_gpg.sh
[...]
read(0, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 8192) = 8192
read(0, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 8192) = 8192
read(0, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 8192) = 8192
read(0, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 8192) = 8192
write(1, "\203}`#s\351:\315\333w\350\350n\23g\1\343\353\1\340\304\273\314;\3538\350\322\3\232\340\321"..., 8192) = 8192
write(1, "\324v,`\310\272\n\rA\266iJ H8\378VV\216t;\237v\356Uv\16\26r\177\373\371"..., 8192) = 8192
write(1, "2\212nK7\301\353\244\354\395\0010\260d\22`\343\230\34\313R\6j\1\t\344\240Q\316\3067'"..., 8192) = 8192
write(1, "qK\211\321&\366H\352.[R\10\39\nu\322KHC\220\396\33@=\272\200U\214\253\277\277\34"..., 8192) = 8192
[...]
```

So gpg seems to use (fixed?) 8k write size.


### What about "lz4"?

```
dd if=/dev/urandom of=/tmp/infile bs=1M count=1
echo 'exec lz4 -c < /tmp/infile > /tmp/out' > /tmp/test_lz4.sh
strace -f sh ./test_lz4.sh
[...]
write(1, "\4\"M\30d`\205\0\0\20\2005;z1\222\177\326\303\236\236\270u\265\237&\7\214L\312\342h"..., 1048576) = 1048576
[...]
write(1, "\22\335v?\20#v\351;u\327\0\0\0\0\216\\\275\16", 19) = 19
[...]
```

Ok, so lz4 writes its complete buffer, all in one go.


## Conclusion:

This all matches with @tobig results above, and it's obvious that acd_cli suffers from every additional write() syscall it has to handle. Maybe that's a fuse issue, but I think it's most probably the multiple layers every write() syscall has to pass: kernel -> fuse -> fusepy(python) -> acd_cli(python). This is probably why acd_cli mounts have such a big block size (524288 from @tobig).

I guess the best workaround for this issue would be to use `| cat >` instead of plain `>`. Still an ugly workaround as on "normal" systems an additional command in the pipe will produce more overhead.

@tobig: Could you please run the `cat` test case above with output file on acd_cli file system? Does it write in chunks of 524288 bytes?

**Comment by tobig** [comment]
On Mon, Aug 22, 2016, at 08:32 PM, Axel Burri wrote:
> I got curious about write sizes of `cat` and `gpg`, this is what I found:

Thank you Alex for this investigation!
 
> ## Conclusion:
> 
> This all matches with @tobig results above, and it's obvious that acd_cli
> suffers from every additional write() syscall it has to handle. Maybe
> that's a fuse issue, but I think it's most probably the multiple layers
> every write() syscall has to pass: kernel -> fuse -> fusepy(python) ->
> acd_cli(python). This is probably why acd_cli mounts have such a big
> block size (524288 from @tobig).

Very interesting observations.
 
> I guess the best workaround for this issue would be to use `| cat >`
> instead of plain `>`. Still an ugly workaround as on "normal" systems an
> additional command in the pipe will produce more overhead.
 
> @tobig: Could you please run the `cat` test case above with output file
> on acd_cli file system? Does it write in chunks of 524288 bytes?

read(3,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
write(1,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
read(3,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
write(1,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
read(3,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
write(1,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
read(3,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
write(1,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
read(3,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
write(1,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
read(3,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
write(1,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
read(3,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
write(1,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
read(3,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072
write(1,
"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,
131072) = 131072

It seems it uses only 128K. I investigated this a little more. On my
system the fstat call returns 4K on the acd filesystem. The reason that
stat -f returns a larger blocksize is that it calls statfs:

statfs("/media/acd/testaa", {f_type=0x65735546, f_bsize=524288,
f_blocks=209715200, f_bfree=209674111, f_bavail=209674111, f_files=0,
f_ffree=0, f_fsid={0, 0}, f_namelen=256, f_frsize=524288}) = 0

As cat always uses at least 128K, plugging cat in between is probably
enough. The difference between 128 and 512 is certainly a lot less than
the one between 8K and the 128K we get with cat. Also, I  would assume
the overhead due to the larger writes is for other filesystems likely
not really a problem.

Best,
Tobias

**Comment by ksobolew** [comment]
Sorry for late reply, for some reason I didn't get notifications about the responses :) The version from devel branch works, though still not much faster. But I'm happy that you're aware of the issue and there is some progress :)

**Comment by tobig** [comment]
If I apply the following patch to acd_cli the blocksize is set correctly and 'cat' uses the blocksize suggested by fuse:

```
diff --git a/acdcli/acd_fuse.py b/acdcli/acd_fuse.py
index 68e026c..2f098c8 100644
--- a/acdcli/acd_fuse.py
+++ b/acdcli/acd_fuse.py
@@ -443,7 +443,8 @@ class ACDFuse(LoggingMixIn, Operations):
 
         times = dict(st_atime=time(),
                      st_mtime=node.modified.timestamp(),
-                     st_ctime=node.created.timestamp())
+                     st_ctime=node.created.timestamp(),
+                     st_blksize=524888)
 
         if node.is_folder:
             return dict(st_mode=stat.S_IFDIR | 0o0777,
```
This means that adding cat into the data stream (and fixing acd_fuse) is probably the cleanest solution.

**Comment by digint** [comment]
For a more correct patch, you should probably use the same as [f_bsize](/yadayada/acd_cli/blob/master/acdcli/acd_fuse.py#L480): `bs = 512 * 1024`

(guess you have a typo, its 524288 not 524888)

**Comment by tobig** [comment]
Right. Interestingly, this mistake does not cause any troubles as at some point the value is rounded down to 524288 as visible in the test_cat.sh trace:

write(1, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 524288) = 524288

In any case, to make this a proper patch the computation of bs should likely be factored out such that we can be certain that the returned values are consistent across the two functions. I will propose a proper patch for acd_cli to improve the block size for acd_cli in btrbk, in case btrbk chooses to use the  'cat' approach to obtain better block sizes.

**Comment by digint** [comment]
I decided to go for a new configuration option which enables "dd" with configurable write size. The discussed "always add cat to pipeline" approach does not seem very robust to me, as it does not enforce any write size, and the implementation might vary in coreutils.

**Comment by tobig** [comment]
OK. Let me know when the patch is available and you want me to test this. Also, I would suggest to enable this by default with a reasonable write size. The overhead of dd is likely to not be measurable and without it many users will be surprised by the low performance.

**Comment by digint** [comment]
Implemented in 961f96833ae8ad26ad29eb7570058b9fcbb081a1

In order to keep things simple, btrbk now always uses "dd", with a block size configurable with the new `raw_target_block_size` option (defaults to 128K).

@ksobolew: Does the default perform well on your SMB mount? Or do you get better results by setting `raw_target_block_size 1M` or higher?

@tobig: Can I ask you to write a little [FAQ entry](/digint/btrbk/blob/master/doc/FAQ.md) on how you managed to get btrbk running with acd_cli? I'm sure many others will profit from it.

**Comment by tobig** [comment]
Thanks. This seems like a good solution. I will let you know when it passes my testing, but I don't expect any issues.

For acd_cli to work out-of-the-box there are still two issue:
1) The file size is not updated immediately by acd_cli
2) When switching networks, the mount becomes unaccessible.

I still work around both of these with local hacks. This probably needs to become a little bit more reliable before we can really recommend this configuration, no?

**Comment by digint** [comment]
> For acd_cli to work out-of-the-box there are still two issue:
1) The file size is not updated immediately by acd_cli
2) When switching networks, the mount becomes unaccessible.
>
> I still work around both of these with local hacks. This probably needs to become a little bit more reliable before we can really recommend this configuration, no?

Yes. At least (1) is still a blocker. Please comment in issue #99 as soon as you have any acceptable patch/workaround for this.

**Comment by ksobolew** [comment]
@digint Yes, 128k works fine (though 1M works just a little finer)
